src/python-common/ceph/deployment/translate.py

```diff
@@ -24,73 +28,158 @@ def __init__(self,
+                continue
+            else:
+                # default use case
+                lvcount["no_crush"].append(path)
+                continue
+
+        return lvcount
+
     def run(self):
         # type: () -> List[str]
         """ Generate ceph-volume commands based on the DriveGroup filters """
-        data_devices = [x.path for x in self.selection.data_devices()]
+
         db_devices = [x.path for x in self.selection.db_devices()]
         wal_devices = [x.path for x in self.selection.wal_devices()]
         journal_devices = [x.path for x in self.selection.journal_devices()]
 
-        if not data_devices:
+        if not self.selection.data_devices():
             return []
 
         cmds: List[str] = []
-        if self.spec.method == 'raw':
-            assert self.spec.objectstore == 'bluestore'
-            # ceph-volume raw prepare only support 1:1 ratio of data to db/wal devices
-            if data_devices and db_devices:
-                if len(data_devices) != len(db_devices):
-                    raise ValueError('Number of data devices must match number of '
-                                     'db devices for raw mode osds')
-            if data_devices and wal_devices:
-                if len(data_devices) != len(wal_devices):
-                    raise ValueError('Number of data devices must match number of '
-                                     'wal devices for raw mode osds')
-            # for raw prepare each data device needs its own prepare command
-            dev_counter = 0
-            while dev_counter < len(data_devices):
-                cmd = "raw prepare --bluestore"
-                cmd += " --data {}".format(data_devices[dev_counter])
-                if db_devices:
-                    cmd += " --block.db {}".format(db_devices[dev_counter])
-                if wal_devices:
-                    cmd += " --block.wal {}".format(wal_devices[dev_counter])
-                cmds.append(cmd)
-                dev_counter += 1
-
-        elif self.spec.objectstore == 'filestore':
+
+        devices = self.prepare_devices()
+        # get the total number of devices provided by the Dict[str, List[str]]
+        devices_count = len(sum(list(devices.values()), []))
+
+        if devices and db_devices:
+            if (devices_count != len(db_devices)) and (self.spec.method == 'raw'):
+                raise ValueError('Number of data devices must match number of '
+                                 'db devices for raw mode osds')
+
+        if devices and wal_devices:
+            if (devices_count != len(wal_devices)) and (self.spec.method == 'raw'):
+                raise ValueError('Number of data devices must match number of '
+                                 'wal devices for raw mode osds')
+
+        # For this use case we don't apply any custom crush_device_classes
+        # Note that filestore is not supported anymore by the DriveGroupSpec
+        if self.spec.objectstore == 'filestore':
```

Comment:
@adk3798 I'm not sure we want to keep L121-L138 (the `filestore` use case). This code is actually never reached, and if filestore is added to the `spec`:

```
ceph config set mgr mgr/cephadm/log_to_cluster_level debug
ceph -W cephadm --watch-debug
```

the spec validation will fail with the following error:

```
    raise DriveGroupValidationError(self.service_id,
ceph.deployment.drive_group.DriveGroupValidationError: Failed to validate OSD spec "testing_drivegroup": filestore is not supported. Must be one of ('bluestore')
```

which is actually expected from [1].
We can leave this block here (and it's out of the for loop), but it's just something we want to clean up (maybe in a follow up change?)

[1] https://github.com/ceph/ceph/blob/main/src/python-common/ceph/deployment/drive_group.py#L328-L331

Reply:
I'll have to check if this is getting used by anything else somehow that could make use of that. Maybe something for a follow up PR, yeah

Reply:
fwiw, filestore is going to be deprecated and removed soon